ffsync/etcd_locker.go

```diff
@@ -0,0 +1,101 @@
+package ffsync
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/featureform/fferr"
+	"github.com/featureform/helpers"
+	"github.com/google/uuid"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/concurrency"
+)
+
+type etcdKey struct {
+	id        string
+	key       string
+	lockMutex *concurrency.Mutex
+}
+
+func (k etcdKey) ID() string {
+	return k.id
+}
+
+func (k etcdKey) Key() string {
+	return k.key
+}
+
+func NewETCDLocker() (Locker, error) {
+	etcdHost := helpers.GetEnv("ETCD_HOST", "localhost")
+	etcdPort := helpers.GetEnv("ETCD_PORT", "2379")
+
+	etcdHostPort := fmt.Sprintf("%s:%s", etcdHost, etcdPort)
+
+	etcdURL := url.URL{
+		Scheme: "http",
+		Host:   etcdHostPort,
+	}
+
+	client, err := clientv3.New(clientv3.Config{
+		Endpoints: []string{etcdURL.String()},
+	})
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to create etcd client: %w", err))
+	}
+
+	session, err := concurrency.NewSession(client)
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to create etcd session: %w", err))
+	}
+
+	return &etcdLocker{
+		client:  client,
+		session: session,
+		ctx:     context.Background(),
+	}, nil
+}
+
+type etcdLocker struct {
+	client  *clientv3.Client
+	session *concurrency.Session
+	ctx     context.Context
+}
+
+func (m *etcdLocker) Lock(key string) (Key, error) {
+	if key == "" {
+		return nil, fferr.NewLockEmptyKeyError()
+	}
+
+	id := uuid.New().String()
+
+	lockMutex := concurrency.NewMutex(m.session, key)
+	if err := lockMutex.Lock(m.ctx); err != nil {
```

Comment:
Deadlock potential, there's no timeout

Reply:
// Grant a lease with a specified TTL
    lease, err := cli.Grant(context.Background(), int64(ttl))
    if err != nil {
        log.Fatalf("Failed to grant a lease: %v", err)
    }

    // Start the keep-alive process for the lease
    leaseKeepAliveChan, err := cli.KeepAlive(context.Background(), lease.ID)
    if err != nil {
        log.Fatalf("Failed to start keep-alive for the lease: %v", err)
    }

    // Go routine to handle keep-alive responses (optional but recommended)
    go func() {
        for {
            select {
            case ka, ok := <-leaseKeepAliveChan:
                if !ok {
                    log.Println("Keep-alive channel closed")
                    return
                }
                log.Printf("Lease keep-alive response received: %+v\n", ka)
            }
        }
    }()

    // Use the lease for a lock or other purposes
    session, err := concurrency.NewSession(cli, concurrency.WithLease(lease.ID))
    if err != nil {
        log.Fatalf("Failed to create a session: %v", err)
    }
    mutex := concurrency.NewMutex(session, lockName)
    
    // Acquire the lock
    if err := mutex.Lock(context.Background()); err != nil {
        log.Fatalf("Failed to acquire lock: %v", err)
    }
    log.Println("Lock acquired")

    // Your service logic here...

    // Release the lock (optional, depends on your use case)
    if err := mutex.Unlock(context.Background()); err != nil {
        log.Fatalf("Failed to release lock: %v", err)
    }
    log.Println("Lock released")
    
    // Clean up: revoke the lease to release the lock immediately when service stops
    if _, err := cli.Revoke(context.Background(), lease.ID); err != nil {
        log.Fatalf("Failed to revoke lease: %v", err)
    }